pkg/reconciler/metric/metric.go

```diff
@@ -49,23 +53,65 @@ func (r *reconciler) Reconcile(ctx context.Context, key string) error {
 	logger := logging.FromContext(ctx)
 	namespace, name, err := cache.SplitMetaNamespaceKey(key)
 	if namespace == "" || err != nil {
-		logger.Errorf("Invalid resource key: %s", key)
+		logger.Error("Invalid resource key: " + key)
 		return nil
 	}
 
-	metric, err := r.metricLister.Metrics(namespace).Get(name)
+	original, err := r.metricLister.Metrics(namespace).Get(name)
 	if apierrs.IsNotFound(err) {
+		// The metric object is gone, so delete the collection.
 		return r.collector.Delete(namespace, name)
 	} else if err != nil {
-		return errors.Wrapf(err, "failed to fetch metric %q", key)
+		return errors.Wrapf(err, "failed to fetch metric "+key)
 	}
 
+	// Don't mess with informer's copy.
+	metric := original.DeepCopy()
+	metric.SetDefaults(ctx)
 	metric.Status.InitializeConditions()
 
-	if err := r.collector.CreateOrUpdate(metric); err != nil {
-		return errors.Wrapf(err, "failed to initiate or update scraping")
+	if err = r.reconcileCollection(ctx, metric); err != nil {
+		logger.Errorw("Error reconciling metric collection", zap.Error(err))
+		r.Recorder.Event(metric, corev1.EventTypeWarning, "InternalError", err.Error())
+	} else {
+		metric.Status.MarkMetricReady()
+	}
+
+	if !equality.Semantic.DeepEqual(original.Status, metric.Status) {
+		// Change of status, need to update the object.
+		if uErr := r.updateStatus(metric); uErr != nil {
+			logger.Warnw("Failed to update metric status", zap.Error(uErr))
+			r.Recorder.Eventf(metric, corev1.EventTypeWarning, "UpdateFailed",
+				"Failed to update metric status: %v", uErr)
+			return uErr
+		}
+		r.Recorder.Eventf(metric, corev1.EventTypeNormal, "Updated", "Successfully updated metric status %s", key)
 	}
+	return err
+}
 
-	metric.Status.MarkMetricReady()
+func (r *reconciler) reconcileCollection(ctx context.Context, metric *v1alpha1.Metric) error {
+	err := r.collector.CreateOrUpdate(metric)
+	if err != nil {
+		// If create or update failes, we won't be able to collect at all.
+		metric.Status.MarkMetricFailed("CollectionFailed", "Failed to reconcile metric collection")
+		return errors.Wrapf(err, "failed to initiate or update scraping")
+	}
 	return nil
 }
+
+func (r *reconciler) updateStatus(m *v1alpha1.Metric) error {
+	ex, err := r.metricLister.Metrics(m.Namespace).Get(m.Name)
```

Comment:
Why do we need to refetch here?

Reply:
Following the pattern elsewhere, but I presume, to make sure things didn't change while we were reconciling.